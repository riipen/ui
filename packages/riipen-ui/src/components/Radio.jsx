import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

import Typography from "./Typography";

class Radio extends React.Component {
  static propTypes = {
    /**
     * If `true`, the component is checked.
     */
    checked: PropTypes.bool,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary", "default"]),

    /**
     * If `true`, the radio will be disabled.
     */
    disabled: PropTypes.bool,

    /**
     * Label text to display for the radio.
     */
    label: PropTypes.string
  };

  static defaultProps = {
    checked: false,
    disabled: false
  };

  static contextType = ThemeContext;

  render() {
    const { checked, classes, color, disabled, label, ...other } = this.props;

    const theme = this.context;

    return (
      <div className={clsx(classes)}>
        <label htmlFor={other.id}>
          <Typography>{label}</Typography>
          <input
            checked={checked}
            disabled={disabled}
            type="radio"
            {...other}
          />
          <span
            className={clsx("checkmark", disabled ? "disabled" : null, color)}
          />
        </label>
        <style jsx>{`
          label {
            cursor: pointer;
            display: block;
            margin-bottom: 12px;
            padding-left: 35px;
            position: relative;
            user-select: none;
          }

          /* Hide the browser's default checkbox */
          label input {
            cursor: pointer;
            opacity: 0;
            position: absolute;
          }

          /* Create a custom checkbox */
          .checkmark {
            background-color: #eee;
            border-radius: 50%;
            border: 1px solid rgba(0, 0, 0, 0.23);
            height: 20px;
            left: 0;
            position: absolute;
            top: 0;
            width: 20px;
          }

          /* When the checkbox is checked */
          label input:checked ~ .checkmark {
            background-color: ${theme.palette.grey[400]};
            border-color: ${theme.palette.grey[400]};
          }

          label input:checked ~ .primary {
            background-color: ${theme.palette.primary.main};
            border-color: ${theme.palette.primary.main};
          }
          label input:checked ~ .secondary {
            background-color: ${theme.palette.secondary.main};
            border-color: ${theme.palette.secondary.main};
          }

          /* Create the checkmark/indicator (hidden when not checked) */
          .checkmark:after {
            content: "";
            display: none;
            position: absolute;
          }

          /* Show the checkmark when checked */
          label input:checked ~ .checkmark:after {
            display: block;
          }

          /* Style the checkmark/indicator */
          label .checkmark:after {
            background: white;
            border-radius: 50%;
            height: 10px;
            left: 5px;
            top: 5px;
            width: 10px;
          }

          .disabled {
            background-color: ${theme.palette.grey[300]};
            border-color: ${theme.palette.grey[300]};
            pointer-events: none;
          }
        `}</style>
      </div>
    );
  }
}

export default Radio;
