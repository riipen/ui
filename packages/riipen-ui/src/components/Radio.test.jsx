import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Radio from "./Radio";

describe("<Radio>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Radio readOnly />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Radio readOnly />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Radio().type.defaultProps;

      const wrapper = mount(<Radio readOnly />);

      expect(wrapper.find("Radio").props().checked).toEqual(
        defaultProps.checked
      );
      expect(wrapper.find("Radio").props().disabled).toEqual(
        defaultProps.disabled
      );
    });
  });

  describe("checked prop", () => {
    it("sets valid custom checked", () => {
      const checked = true;

      const wrapper = mount(<Radio checked={checked} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .find("input")
          .props().checked
      ).toEqual(checked);
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<Radio classes={classes} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-radio"];

      const wrapper = mount(<Radio readOnly />);

      expect(wrapper.find("Radio").props().classes).toEqual(classes.sort());
    });
  });

  describe("color prop", () => {
    it("sets valid custom  color", () => {
      const color = "default";

      const wrapper = mount(<Radio color={color} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .find("Typography")
          .childAt(0)
          .hasClass(color)
      ).toBeTruthy();
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const color = "purple";

      mount(<Radio color={color} readOnly />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("disabled prop", () => {
    it("renders radio as disabled when disabled prop is true", () => {
      const disabled = true;

      const wrapper = mount(<Radio disabled={disabled} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .find("span")
          .hasClass("disabled")
      ).toBeTruthy();
    });

    it("renders radio with color when disabled prop is false", () => {
      const color = "primary";

      const wrapper = mount(<Radio color={color} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .find("span")
          .hasClass(color)
      ).toBeTruthy();
      expect(
        wrapper
          .find("Radio")
          .find("span")
          .hasClass("disabled")
      ).toBeFalsy();
    });
  });

  describe("label prop", () => {
    it("renders custom label", () => {
      const label = "Pizza";

      const wrapper = mount(<Radio label={label} readOnly />);

      expect(
        wrapper
          .find("Radio")
          .find("label")
          .text()
      ).toEqual(label);
    });
  });
});
