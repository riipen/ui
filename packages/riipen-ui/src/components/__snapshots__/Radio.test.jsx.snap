// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio> renders correct snapshot 1`] = `
<withClasses(Radio)
  readOnly={true}
>
  <Radio
    checked={false}
    classes={
      Array [
        "riipen",
        "riipen-radio",
      ]
    }
    disabled={false}
    readOnly={true}
  >
    <div
      className="jsx-23639528 riipen riipen-radio"
    >
      <label
        className="jsx-23639528"
      >
        <input
          checked={false}
          className="jsx-23639528 jsx-835554889"
          disabled={false}
          readOnly={true}
          type="radio"
        />
        <withClasses(Typography)
          classes={
            Array [
              "jsx-835554889",
              undefined,
            ]
          }
        >
          <Typography
            classes={
              Array [
                "riipen",
                "riipen-typography",
                "jsx-835554889",
                undefined,
              ]
            }
            color="inherit"
            display="initial"
            gutter={false}
            mobileBreakpoint="sm"
            textAlign="inherit"
            textTransform="inherit"
            variant="body1"
          >
            <p
              className="jsx-358467856 root color-inherit initial mobile body1 align-inherit transform-inherit riipen riipen-typography jsx-835554889"
            >
              <JSXStyle
                dynamic={
                  Array [
                    "#cccbcb",
                    "#1f71ae",
                    "#f58632",
                  ]
                }
                id="3026427841"
              />
            </p>
            <JSXStyle
              dynamic={
                Array [
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "16px",
                  "24px",
                  "0.00938em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "14px",
                  "20px",
                  "0.01071em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "12px",
                  "16px",
                  "0.01071em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "40px",
                  "56px",
                  "-0.01562em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "30px",
                  "44px",
                  "-0.00833em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "22px",
                  "32px",
                  "0em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "20px",
                  "28px",
                  "-0.00735em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "16px",
                  "24px",
                  "0rem",
                  "#000",
                  "#cccbcb",
                  "#9a9a9a",
                  "#747474",
                  "#424242",
                  "#373737",
                  "#1f71ae",
                  "#f58632",
                  "#3caabb",
                  "#4cc88d",
                  "#e95353",
                  "#febb46",
                  "#fff",
                  700,
                  300,
                  500,
                  400,
                  480,
                  "14px",
                  "20px",
                  "12px",
                  "16px",
                  "11px",
                  "16px",
                  "30px",
                  "44px",
                  "22px",
                  "32px",
                  "18px",
                  "28px",
                  "16px",
                  "24px",
                  "14px",
                  "20px",
                ]
              }
              id="3367936398"
            />
          </Typography>
        </withClasses(Typography)>
        <span
          className="jsx-23639528 checkmark"
        />
      </label>
      <JSXStyle
        dynamic={
          Array [
            "#747474",
            20,
            "#fff",
            "#cccbcb",
            "#1f71ae",
            "#f58632",
            "#cccbcb",
            "#1f71ae",
            "#f58632",
            "#cccbcb",
          ]
        }
        id="1070945688"
      />
    </div>
  </Radio>
</withClasses(Radio)>
`;
